File: Medium/1743. Restore the Array From Adjacent Pairs/1743. Restore the Array From Adjacent Pairs.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public int[] restoreArray(int[][] adjacentPairs) {
        Map<Integer, List<Integer>> adjacencyMap = new HashMap<>();

        // Build the adjacency map
        for (int[] pair : adjacentPairs) {
            adjacencyMap.computeIfAbsent(pair[0], k -> new ArrayList<>()).add(pair[1]);
            adjacencyMap.computeIfAbsent(pair[1], k -> new ArrayList<>()).add(pair[0]);
        }

        int n = adjacentPairs.length + 1;
        int[] result = new int[n];
        Set<Integer> visited = new HashSet<>();

        // Find the starting point
        int start = 0;
        for (Map.Entry<Integer, List<Integer>> entry : adjacencyMap.entrySet()) {
            if (entry.getValue().size() == 1) {
                start = entry.getKey();
                break;
            }
        }

        // Reconstruct the array
        result[0] = start;
        result[1] = adjacencyMap.get(start).get(0);

        visited.add(start);
        visited.add(result[1]);

        for (int i = 2; i < n; i++) {
            List<Integer> neighbors = adjacencyMap.get(result[i - 1]);
            result[i] = neighbors.get(0) == result[i - 2] ? neighbors.get(1) : neighbors.get(0);

            visited.add(result[i]);
        }

        return result;
    }
}

File: Medium/1743. Restore the Array From Adjacent Pairs/1743. Restore the Array From Adjacent Pairs.js
/**
 * @param {number[][]} adjacentPairs
 * @return {number[]}
 */
var restoreArray = function (adjacentPairs) {
    const adjacencyMap = new Map();

    // Build the adjacency map
    for (const pair of adjacentPairs) {
        if (!adjacencyMap.has(pair[0])) {
            adjacencyMap.set(pair[0], []);
        }
        if (!adjacencyMap.has(pair[1])) {
            adjacencyMap.set(pair[1], []);
        }

        adjacencyMap.get(pair[0]).push(pair[1]);
        adjacencyMap.get(pair[1]).push(pair[0]);
    }

    const n = adjacentPairs.length + 1;
    const result = new Array(n);
    const visited = new Set();

    // Find the starting point
    let start = 0;
    for (const [key, value] of adjacencyMap.entries()) {
        if (value.length === 1) {
            start = key;
            break;
        }
    }

    // Reconstruct the array
    result[0] = start;
    result[1] = adjacencyMap.get(start)[0];

    visited.add(start);
    visited.add(result[1]);

    for (let i = 2; i < n; i++) {
        const neighbors = adjacencyMap.get(result[i - 1]);
        result[i] =
            neighbors[0] === result[i - 2] ? neighbors[1] : neighbors[0];

        visited.add(result[i]);
    }

    return result;
};
